infra/terraform/modules/kubernetes_cluster/bin/kops-specs/generate_specs.py

```diff
@@ -0,0 +1,211 @@
+parser.add_argument('--state-bucket', nargs='?', required=True)
+parser.add_argument('--kubernetes-version', nargs='?', required=True)
+parser.add_argument('--dns-zone', nargs='?', required=True)
+parser.add_argument('--zones', nargs='?', required=True)
+parser.add_argument('--network-cidr', nargs='?', required=True)
+parser.add_argument('--node-instance-type', nargs='?', required=True)
+parser.add_argument('--node-count', type=int, nargs='?', required=True)
+parser.add_argument('--node-volume-size', type=int, nargs='?', required=True)
+parser.add_argument('--ami-name', nargs='?', required=True)
+parser.add_argument('--bastion-instance-type', nargs='?', required=True)
+parser.add_argument('--bastion-count', type=int, nargs='?', required=True)
+parser.add_argument('--master-instance-type', nargs='?', required=True)
+parser.add_argument('--public-subnet-zones', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--private-subnet-zones', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--public-subnet-cidrs', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--private-subnet-cidrs', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--nat-gateway-subnets', type=json.loads, nargs='?',
+                    required=True)
+
+
+args = parser.parse_args()
+
+if not os.path.exists(args.out_path):
+    os.makedirs(args.out_path)
+
+def template(fname):
+    return os.path.join(args.template_path, fname)
+
+def outfile(fname):
+    return os.path.join(args.out_path, fname)
+
+
+zones = sorted(args.zones.split(','))
+
+nat_gateway_zones = {
+    args.public_subnet_zones[k] : v
+    for k, v in args.nat_gateway_subnets.items()
+}
+
+subnets = {
+    'public': [],
+    'private': []
+}
+for kind in ['private', 'public']:
+    cidrs = vars(args)["{}_subnet_cidrs".format(kind)]
+    subnet_zones = vars(args)["{}_subnet_zones".format(kind)]
+
+    for subnet_id, zone in sorted(subnet_zones.items(), key=lambda i: i[1]):
+        subnet = {
+            'id': subnet_id,
+            'zone': zone,
+            'cidr': cidrs[subnet_id],
+            'type': 'Private' if kind == 'private' else 'Utility',
+            'name': zone if kind == 'private' else 'utility-{}'.format(zone)
+        }
+
+        if kind == 'private':
+            subnet['egress'] = nat_gateway_zones[zone]
+
+        subnets[kind].append(subnet)
+
+# Cluster spec
+with open(template('cluster.tpl.yml'), 'r') as stream:
+    try:
+        cluster = yaml.load(stream)
+    except yaml.YAMLError as exc:
+        raise(exc)
+
+    cluster['metadata'] = {'name': args.cluster_name}
+
+    cluster['spec'].update({
+        'etcdClusters': [
+            {
+                'name': name,
+                'etcdMembers': [
+                    {
+                        'instanceGroup': 'master-{}'.format(zone),
+                        'name': zone[-1]
```

Comment:
This could use some clarification - why are you taking the last element of `zone` here? Perhaps this should be extracted into a function which has a self-documenting name.